### DISCUSSION TITLE
Disk performance and recommended setup

### DISCUSSION THREAD

**Discussion by meb427** [open-meteo/open-meteo#504] [General]
Hi community, Patrick,

Can you share more about the configuration you are using and best recommended performance setup? Disks used? To get a clear idea what to expect in performance ratio?

On my side I notice despite having a Crucial disk in NVME, it's still having a lot of busy IO in CPU.

Thanks

**Comment by patrick-zippenfenig** [comment]
Hi,

high disk usage is normal while downloading and converting weather models. Because weather models need to be converted to a time-series, a lot of small random IO reads are required. In general this can be improved with more system memory. 

The API part requires a lot of random IO as well. Due to prefetching SSDs will be fully utilised with high queue depths, but in most cases, the CPU will be a bottleneck before a NVMe disk becomes a bottleneck.

The Open-Meteo servers use Samsung Enterprise SSDs. I am not too focused on performance, because downloading and database updates are done and separate servers. Ideally, I would like to publicly redistribute the converted database files, but due to cost of bandwidth/traffic this is not possible.

Do you notice any bottlenecks? How are IOPS, disk bandwidth, CPU and memory utilisation while seeing busy IO?

**Comment by meb427** [comment]
I had tested 2 configs for mainly processing and download a Ryzen 9 and an Atom Alder Lake, as expected here the Atom took the hit and is the bottle neck. But on the Ryzen I got a lot of busy IO and the disk (tested 2 different NVME not entreprise) got at 100%. I will try a third config based on Ryzen 7. Could share more infos in grafana if you are intersted in private.
Curious also about the CPU you use for processing data here. I'm still heavily testing. 

I've also thought about sharing the DB, because historical data download is a total pain. We could use for that R2 https://developers.cloudflare.com/r2/get-started/ cheaper and if I'm not wrong sharing public data on AWS is free https://aws.amazon.com/opendata/ .

I'm thinking about using multi node too. Will test it too.

**Reply by patrick-zippenfenig** [reply]
100% disk utilisation can be ok during the last conversion step. It is pretty read-IO heavy.

There is progress in sharing the converted Open-Meteo database files as open-data, but it will take some time until it is available.

**Reply by meb427** [reply]
As you said it. The conversion is heavy task but once we don't have option to set different storage that will hammer same disk with Read and write and thus degrade performance on both sides.
I'm currently cheating with Symbolic links as I dispatched that way the load between disk and I think I will add a third disk only for download and so reading and test again. This way I'm faster as I feel the CPU is not fully used as I do all the converion on one Ryzen 9.

I started too logging the time it takes to do some operations to identify the heavy jobs. I wave now a shell wrapper for the cron that produce and plan to push these metrics to pushgateway/prometheus.

download icon-icon_upper-levelgt500 last ran and finished at 2023-11-15 09:42:45. Duration: 3944 seconds.
download-metno-nordic_pp last ran and finished at 2023-11-15 10:36:41. Duration: 580 seconds.
download icon-eu-icon-eu_model-level last ran and finished at 2023-11-15 09:21:54. Duration: 2753 seconds.
download icon-icon_model-level last ran and finished at 2023-11-15 09:45:20. Duration: 4099 seconds.
generic_job-generic_log last ran and finished at 2023-11-14 10:23:07. Duration: 0 seconds.
download icon-eps-icon-eps last ran and finished at 2023-11-15 04:52:40. Duration: 8139 seconds.
download-gfs-hrrr_conus_15min last ran and finished at 2023-11-15 10:29:39. Duration: 2078 seconds.
download-gem-gem_hrdps_continental last ran and finished at 2023-11-15 10:24:21. Duration: 4640 seconds.
download-ecmwf-ecmwf last ran and finished at 2023-11-15 07:56:02. Duration: 661 seconds.
download icon-icon last ran and finished at 2023-11-15 09:25:24. Duration: 2903 seconds.
download icon-icon_upper-levellte500 last ran and finished at 2023-11-15 09:56:57. Duration: 4796 seconds.
download icon-d2-icon-d2 last ran and finished at 2023-11-15 10:30:19. Duration: 2778 seconds.
download icon-eu-icon-eu last ran and finished at 2023-11-15 09:41:52. Duration: 3951 seconds.
download-iconwave-iconwave_gwam last ran and finished at 2023-11-15 03:56:39. Duration: 998 seconds.
download-gfs-gfs05_ens-second-flush last ran and finished at 2023-11-14 23:55:01. Duration: 0 seconds.
download icon-eu-icon-eu_upper-level last ran and finished at 2023-11-15 09:47:57. Duration: 4316 seconds.
download-gfs-gfs025_upper-level last ran and finished at 2023-11-13 12:08:24. Duration: 8903 seconds.
download-meteofrance-arpege_world_upper-level last ran and finished at 2023-11-15 09:00:02. Duration: 1 seconds.
download icon-d2-icon-d2_upper-level last ran and finished at 2023-11-15 10:14:19. Duration: 1818 seconds.
download-ecmwf-ifs04_ensemble last ran and finished at 2023-11-15 08:32:24. Duration: 2843 seconds.
download-meteofrance-arome_france_upper-level last ran and finished at 2023-11-15 10:26:32. Duration: 8791 seconds.
download icon-d2-eps-icon-d2-eps last ran and finished at 2023-11-15 10:30:03. Duration: 2 seconds.
download icon-eu-eps-icon-eu-eps last ran and finished at 2023-11-15 09:20:50. Duration: 2629 seconds.
download-gfs-gfs05_ens last ran and finished at 2023-11-15 09:40:02. Duration: 1 seconds.
download-meteofrance-arpege_europe_upper-level last ran and finished at 2023-11-15 09:00:02. Duration: 1 seconds.
download-gem-gem_global_ensemble last ran and finished at 2023-11-15 05:11:17. Duration: 1576 seconds.
download-iconwave-iconwave_ewam last ran and finished at 2023-11-15 03:46:08. Duration: 967 seconds.
download-meteofrance-arpege_europe last ran and finished at 2023-11-15 09:00:02. Duration: 1 seconds.
download-gem-gem_regional last ran and finished at 2023-11-15 10:05:55. Duration: 4734 seconds.
download-gem-gem_global last ran and finished at 2023-11-15 05:39:13. Duration: 7212 seconds.
download-meteofrance-arpege_world last ran and finished at 2023-11-15 09:00:02. Duration: 1 seconds.
download-seasonal-forecast-seasonal_ncep last ran and finished at 2023-11-15 05:23:18. Duration: 197 seconds.
download-gfs-hrrr_conus_upper-level last ran and finished at 2023-11-15 10:34:51. Duration: 2390 seconds.
download-gfs-hrrr_conus last ran and finished at 2023-11-15 10:30:38. Duration: 2137 seconds.
download-meteofrance-arome_france_hd last ran and finished at 2023-11-15 08:00:02. Duration: 1 seconds.
download-meteofrance-arome_france last ran and finished at 2023-11-15 10:27:24. Duration: 8843 seconds.

I got this combined with nice to lower priority to get far better performance and get out of the busy io wait on CPU.
![image](https://github.com/open-meteo/open-meteo/assets/128685138/9a4a3e6d-dcc6-429c-8a77-1ddd56745f84)

**Reply by patrick-zippenfenig** [reply]
Interesting. How did the conversion speed change after changing to `nice`? I am confused why even `user` cpu utilisation is almost 0.

One potential knob to tune on Open-Meteo is the write-sync ratio. I am calling `sync` after [writing 32 MB](https://github.com/open-meteo/open-meteo/blob/9c82c72bfd25b8dc5fdfeb46a03c9c5f74ed99c2/Sources/SwiftPFor2D/SwiftPFor2D.swift#L205). This makes sure that data is actually written to disk and does not overload the kernel with high dirty page ratios. In certain situations high dirty-ratios can cause stalls.

**Reply by meb427** [reply]
These are numbers post nice. 
Unfortunatly I didn't had some before tweaking "nice" but I can revert to compare over 1 or 2 days, not an issue for me to do those tests.
But since I added nice I saw quicly less Busy IoWait on CPU that was blocking the server and spilling over all the services there.

This CPU is a Ryzen 9 6900HX 8 Cores 16 Threads. And yes surprised too, even before using nice the CPU didn't spike to 100%. Biggest issue is disk it seems.
